transport: log network reconnects with same peer process

schase-es opened the pull request:
ClusterConnectionManager now caches the previous ephemeralId (created on process-start) of peer nodes on disconnect in a connection history table. On reconnect, when a peer has the same ephemeralId as it did previously, this is logged to indicate a network failure. The connectionHistory is trimmed to the current set of peers by NodeConnectionsService.

schase-es committed:
transport: log network reconnects with same peer process

ClusterConnectionManager now caches the previous ephemeralId (created on
process-start) of peer nodes on disconnect in a connection history table. On
reconnect, when a peer has the same ephemeralId as it did previously, this is
logged to indicate a network failure. The connectionHistory is trimmed to the
current set of peers by NodeConnectionsService.

elasticsearchmachine commented:
Pinging @elastic/es-distributed-coordination (Team:Distributed Coordination)

schase-es commented:
I wasn't able to find a way to test the ClusterConnectionManager's connectionHistory table when integrated through the NodeConnectionsService.

Commit:
[CI] Auto commit changes from spotless

nicktindall commented on server/src/main/java/org/elasticsearch/transport/ClusterConnectionManager.java:
```diff
@@ -235,25 +266,65 @@ private void connectToNodeOrRetry(
+                                    public void onFailure(Exception e) {
+                                        final NodeConnectionHistory hist = new NodeConnectionHistory(node.getEphemeralId(), e);
+                                        nodeHistory.put(conn.getNode().getId(), hist);
+                                    }
```
Do we want to store the connection history even when `conn.hasReferences() == false` ? I'm not 100% familiar with this code, but I wonder if we might get the occasional ungraceful disconnect after we've released all our references?

I guess in that case we would eventually discard the entry via `retainConnectionHistory` anyway.

Do we need to be careful with the timing of calls to `retainConnectionHistory` versus the these close handlers firing? I guess any entries that are added after a purge would not survive subsequent purges.

nicktindall commented on server/src/main/java/org/elasticsearch/transport/ClusterConnectionManager.java:
```diff
@@ -235,25 +266,65 @@ private void connectToNodeOrRetry(
+                                                node.descriptionWithoutAttributes(),
+                                                e,
+                                                ReferenceDocs.NETWORK_DISCONNECT_TROUBLESHOOTING
+                                            );
```
It looks like previously we would only have logged at debug level in this scenario? unless I'm reading it wrong. I'm not sure how interesting this case is (as we were disconnecting from the node anyway)?

nicktindall commented on server/src/test/java/org/elasticsearch/transport/ClusterConnectionManagerTests.java:
```diff
@@ -207,9 +236,64 @@ public void testDisconnectLogging() {
+            assertTrue("recent disconnects should be listed", connectionManager.connectionHistorySize() == 2);
+
+            connectionManager.retainConnectionHistory(Collections.emptyList());
+            assertTrue("connection history should be emptied", connectionManager.connectionHistorySize() == 0);
```
I wonder if it would be better to expose a read-only copy of the map for testing this, that would allow us to assert that the correct IDs were present?

nicktindall reviewed (commented):
Looking good, just a few questions and minor comments.

DaveCTurner reviewed (commented):
I think `ClusterConnectionManager` isn't quite the right place to do this - the job of this connection manager is to look after all node-to-node connections including ones used for discovery and remote cluster connections too. There are situations where we might close and re-establish these kinds of connection without either end restarting without that being a problem worthy of logging.

`NodeConnectionsService` is the class that knows about connections to nodes in the cluster. I'd rather we implemented the logging about unexpected reconnects there. That does raise some difficulties about how to expose the exception that closed the connection, if such an exception exists. I did say that this bit would be tricky :grin: Nonetheless I'd rather we got the logging to happen in the right place first and then we can think about the plumbing needed to achieve this extra detail.

nicktindall commented on server/src/main/java/org/elasticsearch/transport/ClusterConnectionManager.java:
```diff
@@ -226,6 +234,29 @@ private void connectToNodeOrRetry(
+                                        """
+                                            transport connection reopened to node with same ephemeralId [{}]""",
+                                        node.descriptionWithoutAttributes()
+                                    );
```
Oh that's right, the ephemeral IDs are equal here.

schase-es committed:
Addressed review feedback:
- moved test out of ESLoggingHandlerIt into a separate ClusterConnectionManagerIntegTests file
- moved connection history into NodeConnectionsService, and adopted a consistency scheme
- rewrote re-connection log message to include duration
- changed log level of local disconnect with exception to debug

DaveCTurner commented on server/src/main/java/org/elasticsearch/transport/ClusterConnectionManager.java:
```diff
@@ -235,25 +235,48 @@ private void connectToNodeOrRetry(
+                                    @Override
+                                    public void onFailure(Exception e) {
+                                        if (conn.hasReferences()) {
+                                            logger.warn(
+                                                """
+                                                    transport connection to [{}] closed by remote with exception [{}]; \
+                                                    if unexpected, see [{}] for troubleshooting guidance""",
```
I think this isn't guaranteed to be a `WARN` worthy event - if the node shut down then we might get a `Connection reset` or similar but that's not something that needs action, and we do log those exceptions elsewhere. On reflection I'd rather leave the logging in `ClusterConnectionManager` alone in this PR and just look at the new logs from the `NodeConnectionsService`.

DaveCTurner commented on server/src/test/java/org/elasticsearch/transport/ClusterConnectionManagerIntegTests.java:
```diff
@@ -0,0 +1,50 @@
+import org.elasticsearch.test.junit.annotations.TestLogging;
+
+@ESIntegTestCase.ClusterScope(numDataNodes = 2, scope = ESIntegTestCase.Scope.TEST)
+public class ClusterConnectionManagerIntegTests extends ESIntegTestCase {
```
nit: `ESIntegTestCase` tests should have names ending in `IT` and be in the `internalClusterTest` source set. But as mentioned in my previous comment we probably don't want to change this here.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java:
```diff
@@ -347,4 +357,113 @@ public String toString() {
         }
     }
+
+    private class ConnectionHistory {
```
Yeah I like the look of this. Maybe `ConnectionHistory implements TransportConnectionListener` rather than having another layer of indirection?

Also this needs to be covered in `NodeConnectionsServiceTests`.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java:
```diff
@@ -347,4 +357,113 @@ public String toString() {
+         * Each node in the cluster always has a nodeHistory entry that is either the dummy value or a connection history record. This
+         * allows node disconnect callbacks to discard their entry if the disconnect occurred because of a change in cluster state.
+         */
+        private final NodeConnectionHistory dummy = new NodeConnectionHistory("", 0, null);
```
Can be `static` I think, it's a global constant. We tend to name global constants in `SHOUTY_SNAKE_CASE` reflecting their meaning, so here I'd suggest `CONNECTED` or `CONNECTED_MARKER` or something like that. This way you get to say `nodeConnectionHistory != CONNECTED_MARKER` below which makes it clearer to the reader what this predicate means.

nicktindall replied:
nit: also looks like the javadoc is for the `nodeHistory` field

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java:
```diff
@@ -347,4 +357,113 @@ public String toString() {
+                                    "reopened transport connection to node [%s] "
+                                        + "which disconnected exceptionally [%dms] ago but did not "
+                                        + "restart, so the disconnection is unexpected; "
+                                        + "if unexpected, see [{}] for troubleshooting guidance",
```
No need for `if unexpected` here, I think the point is that this situation is _always_ unexpected.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java:
```diff
@@ -347,4 +357,113 @@ public String toString() {
+                                        + "restart, so the disconnection is unexpected; "
+                                        + "if unexpected, see [{}] for troubleshooting guidance",
+                                    node.descriptionWithoutAttributes(),
+                                    nodeConnectionHistory.disconnectTime,
```
This'll show the absolute disconnect time in milliseconds (i.e. since 1970) whereas I think we want to see the duration between the disconnect and the current time.

nicktindall commented on server/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java:
```diff
@@ -347,4 +357,113 @@ public String toString() {
+
+        void reserveConnectionHistoryForNodes(DiscoveryNodes nodes) {
+            for (DiscoveryNode node : nodes) {
+                nodeHistory.put(node.getId(), dummy);
```
This might need to be `putIfAbsent` so we don't over-write any actual current `NodeConnectionHistory` entries right?

schase-es replied:
I'm not sure. My read was these two calls would come from cluster state changing to add or remove nodes from this table. Inclusion is controlled by these calls, which unconditionally add or remove entries. The close callback has to be careful to check if it has an entry that's valid: this protects against long-running callbacks inserting garbage into the table.

DaveCTurner replied:
The `DiscoveryNodes` passed to `connectToNodes` contains all the nodes in the cluster, including any existing ones, so if there's a node which already exists in the cluster, and is currently disconnected, then it will have an entry in `nodeHistory` which isn't `dummy` that this line will overwrite on any cluster state update. So yeah I think `putIfAbsent` is what we want here.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java:
```diff
@@ -347,4 +357,113 @@ public String toString() {
+                public void onNodeConnected(DiscoveryNode node, Transport.Connection connection) {
+                    // log case where the remote node has same ephemeralId as its previous connection
+                    // (the network was disrupted, but not the remote process)
+                    NodeConnectionHistory nodeConnectionHistory = nodeHistory.get(node.getId());
+                    if (nodeConnectionHistory != null) {
+                        nodeHistory.replace(node.getId(), nodeConnectionHistory, dummy);
+                    }
```
This looks a little racy, although in practice I think it's fine because `ClusterConnectionManager` protects against opening multiple connections to the same node concurrently. Still, if we did all this (including the logging) within a `nodeHistory.compute(node.getId, ...)` then there'd obviously be no races.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java:
```diff
@@ -347,4 +357,113 @@ public String toString() {
+        void removeConnectionHistoryForNodes(Set<DiscoveryNode> nodes) {
+            final int startSize = nodeHistory.size();
+            for (DiscoveryNode node : nodes) {
+                nodeHistory.remove(node.getId());
```
There's kind of an implicit invariant here that `org.elasticsearch.cluster.NodeConnectionsService.ConnectionHistory#nodeHistory` and `org.elasticsearch.cluster.NodeConnectionsService#targetsByNode` have the same keys. At the very least we should be able to assert this. I also wonder if we should be calling `nodeHistory.retainAll()` to make it super-clear that we are keeping these keysets aligned.

But then that got me thinking, maybe we should be tracking the connection history of each target node in `ConnectionTarget` rather than trying to maintain two parallel maps. Could that work?

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java:
```diff
@@ -347,4 +357,113 @@ public String toString() {
+                }
+
+                @Override
+                public void onNodeDisconnected(DiscoveryNode node, Transport.Connection connection) {
```
I just spotted we're already executing this in a close-listener, but one that runs under `ActionListener.running(...)` so it drops the exception. I think it'd be nicer to adjust this callback to take a `@Nullable Exception e` parameter rather than having to add a second close listener just to pick up the exception as done here.

schase-es committed:
Addressed 2nd round review feedback:
- consolidated ConnectionHistory into ConnectionTarget, protected with the service mutex
- added logging test for reconnection with and without exception
- grew TransportConnectionListener onNodeDisconnected to include a nullable exception
- reverted ClusterConnectionManager tests and logging

DaveCTurner commented:
> I haven't been able to track down where the `ephemeralId` gets set

It's created (randomly) once during node startup, see `org.elasticsearch.cluster.node.DiscoveryNode#createLocal`. It is admittedly rather hard to track that down.

> it's not possible for the discovery node view that the `NodeConnectionsService` has when it establishes the connection to have the correct `ephemeralId`

The `NodeConnectionsService` sees the `DiscoveryNode` only after it is added to the `ClusterState` by the cluster-joining process, which means it has all the right details (they're passed over the wire from the joining node to the master in `org.elasticsearch.cluster.coordination.JoinRequest#sourceNode`).

DaveCTurner reviewed (commented):
Yep I think this is going to work. A few superficial comments but otherwise this looks ready to start working on some testing.

schase-es commented on server/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java:
```diff
@@ -347,4 +358,73 @@ public String toString() {
     }
+
+    private class ConnectionHistoryListener implements TransportConnectionListener {
+        /**
+         * Receives connection/disconnection events from the transport, and records in per-node ConnectionHistory
+         * structures for logging network issues. ConnectionHistory records are stored in ConnectionTargets.
```
Argh, my python docstring habits are getting the better of me :)

schase-es commented:
Thanks also for the notes about ephemeral ids. I understood the lifecycle aspect, but was confused about a peer node hears about it. In my mind, I was confused about how it knows the ephemeral id before opening a connection, because I expected you'd need the connection to know the ephemeral id. The chicken-egg aspect of the cluster state now seems obvious -- it wouldn't be in cluster state if the peer hadn't heard from it...

schase-es commented on server/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java:
```diff
@@ -79,12 +83,14 @@ public class NodeConnectionsService extends AbstractLifecycleComponent {
         this.threadPool = threadPool;
         this.transportService = transportService;
         this.reconnectInterval = NodeConnectionsService.CLUSTER_NODE_RECONNECT_INTERVAL_SETTING.get(settings);
+        this.connectionHistoryListener = new ConnectionHistoryListener();
```
I gave this a try, and do prefer your suggestion.

It turned out that all the test failures are from the roughly 1000 tests that instantiate `NodeConnectionsService` with a `null` `TransportService`.

I went for a different solution that calls into `addConnectionListener` during `doStart`, removes it `doStop`, and does need to store the listener as a field to support this.

DaveCTurner replied:
:+1: tho there's no need to remove the listener in `doStop`

schase-es committed:
Addressed next round of feedback:
- updated ConnectionChangeListener constructor and moved registration to service.doStart()
- renamed DisconnectionHistory record and ConnectionChangeListener
- fixed up DiscoveryNode vs. Transport.Connection.getNode() confusion (these are the same)
- fixed log formatting
- edited TransportConnectionListener interface to take a nullable closeException instead of the Transport.Connection
- completed test of DisconnectionHistory, at init, post-connection, post-disconnection, and post-reconnection
- moved docs for ConnectionChangeListener, and added docs for DisconnectionHistory

DaveCTurner commented on server/src/main/java/org/elasticsearch/transport/TransportConnectionListener.java:
```diff
@@ -38,5 +39,5 @@ default void onNodeConnected(DiscoveryNode node, Transport.Connection connection
      * Called once a node connection is closed and unregistered.
      */
-    default void onNodeDisconnected(DiscoveryNode node, Transport.Connection connection) {}
+    default void onNodeDisconnected(DiscoveryNode node, @Nullable Exception closeException) {}
```
I [mentioned earlier](https://github.com/elastic/elasticsearch/pull/128415#discussion_r2117382024) that we _could_ consider pulling this API change out to a separate PR. As things stand I now think we should definitely do that - it's a simple refactoring (needs no test changes) and will make this change much more focussed.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java:
```diff
@@ -209,12 +214,36 @@ public void reconnectToNodes(DiscoveryNodes discoveryNodes, Runnable onCompletio
     }
 
-    private class ConnectionTarget {
+    // exposed for testing
+    protected ConnectionTarget connectionTargetForNode(DiscoveryNode node) {
```
I don't think we need to expose the whole `ConnectionTarget` out to tests - we could just allow access to the `DisconnectionHistory` for a node and keep the `ConnectionTarget` class private.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java:
```diff
@@ -209,12 +214,36 @@ public void reconnectToNodes(DiscoveryNodes discoveryNodes, Runnable onCompletio
         private final AtomicReference<Releasable> connectionRef = new AtomicReference<>();
 
+        // access is synchronized by the service mutex
+        protected DisconnectionHistory disconnectionHistory = null;
```
nit: suggest marking this nullable (and describing what the null value means)

```suggestion
        @Nullable // if node is connected
        protected DisconnectionHistory disconnectionHistory = null;
```

DaveCTurner commented on server/src/test/java/org/elasticsearch/cluster/NodeConnectionsServiceTests.java:
```diff
@@ -245,13 +249,111 @@ public String toString() {
+        final DeterministicTaskQueue deterministicTaskQueue = new DeterministicTaskQueue();
+        final ThreadPool threadPool = deterministicTaskQueue.getThreadPool();
+
+        MockTransport transport = new MockTransport(deterministicTaskQueue.getThreadPool());
+        TestTransportService transportService = new TestTransportService(transport, deterministicTaskQueue.getThreadPool());
```
Should probably use `threadPool` here too rather than creating distinct threadpools for each service.

DaveCTurner commented on server/src/test/java/org/elasticsearch/cluster/NodeConnectionsServiceTests.java:
```diff
@@ -245,13 +249,111 @@ public String toString() {
 
+    public void testDisconnectionHistory() {
+        final Settings.Builder settings = Settings.builder();
+        settings.put(CLUSTER_NODE_RECONNECT_INTERVAL_SETTING.getKey(), "100ms");
```
We're using a `DeterministicTaskQueue` so we are simulating the passage of time, no need to set a short retry interval here. I'd suggest dropping this and just using the default (via `CLUSTER_NODE_RECONNECT_INTERVAL_SETTING.get(Settings.EMPTY).millis()`) below. We could also randomly pick a different value:

```java
        final long reconnectIntervalMillis;
        if (randomBoolean()) {
            reconnectIntervalMillis = randomLongBetween(1, 100000);
            settings.put(CLUSTER_NODE_RECONNECT_INTERVAL_SETTING.getKey(), TimeValue.timeValueMillis(reconnectIntervalMillis));
        } else {
            reconnectIntervalMillis = CLUSTER_NODE_RECONNECT_INTERVAL_SETTING.get(Settings.EMPTY).millis();
        }
```

It's not really the point of this test tho, so I think the default would be fine.

DaveCTurner commented on server/src/test/java/org/elasticsearch/cluster/NodeConnectionsServiceTests.java:
```diff
@@ -245,13 +249,111 @@ public String toString() {
+                    Level.WARN,
+                    "reopened transport connection to node ["
+                        + exceptionalClose.descriptionWithoutAttributes()
+                        + "] which disconnected exceptionally [*ms] ago "
```
We control the passage of time so we can assert that the reconnect happens exactly one reconnect interval later:

```suggestion
                        + "] which disconnected exceptionally ["
                        + CLUSTER_NODE_RECONNECT_INTERVAL_SETTING.get(Settings.EMPTY).millis()
                        + "ms] ago "
```

DaveCTurner commented on server/src/test/java/org/elasticsearch/cluster/NodeConnectionsServiceTests.java:
```diff
@@ -245,13 +249,111 @@ public String toString() {
+            assertDisconnectionHistoryDetails(service, threadPool, gracefulClose, null);
+            assertDisconnectionHistoryDetails(service, threadPool, exceptionalClose, RuntimeException.class);
+
+            runTasksUntil(deterministicTaskQueue, 200);
```
I'd prefer this to be `deterministicTaskQueue.getCurrentTimeMillis() + ${RECONNECT_INTERVAL}` - it just happens that today we start the clock at zero.

schase-es replied:
Oh I see -- I completely misread that as a time duration

schase-es committed:
Merge remote-tracking branch 'origin/main' into ES-11448_log-network-disruption-reconnects_rebase-after-refactoring

schase-es committed:
Addressed next set of review feedback:
- expose only DisconnectionHistory instead of ConnectionTarget as protected
- Nullable annotation on ConnectionTarget's DisconnectionHistory field
- log seconds and milliseconds since last connect, as #s/#ms
- in test, re-use thread pool, narrowed try-block for log checking, used default reconnection period, and updated logs to test for time formatting

Commit:
[CI] Auto commit changes from spotless

DaveCTurner reviewed (commented):
Looks great, just one comment about the human-readable times in the message

schase-es commented on server/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java:
```diff
@@ -345,4 +379,70 @@ public String toString() {
+                        () -> format(
+                            """
+                                reopened transport connection to node [%s] \
+                                which disconnected exceptionally [%ds/%dms] ago but did not \
```
Oh I see I got lost in understanding what TimeValue puts out. Thanks!

schase-es committed:
Addressed review feedback for TimeValue instead of just seconds

DaveCTurner commented on server/src/test/java/org/elasticsearch/cluster/NodeConnectionsServiceTests.java:
```diff
@@ -246,6 +249,110 @@ public String toString() {
+    public void testDisconnectionHistory() {
+        final DeterministicTaskQueue deterministicTaskQueue = new DeterministicTaskQueue();
+        final ThreadPool threadPool = deterministicTaskQueue.getThreadPool();
+        final long reconnectIntervalMillis = CLUSTER_NODE_RECONNECT_INTERVAL_SETTING.get(Settings.EMPTY).millis();
+        final TimeValue reconnectIntervalTimeValue = TimeValue.timeValueMillis(reconnectIntervalMillis);
```
nit: suggest doing these in the opposite order:

```suggestion
        final TimeValue reconnectIntervalTimeValue = CLUSTER_NODE_RECONNECT_INTERVAL_SETTING.get(Settings.EMPTY);
        final long reconnectIntervalMillis = reconnectIntervalTimeValue.millis();
```

DaveCTurner reviewed (approved):
LGTM great stuff

ywangd commented on server/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java:
```diff
@@ -209,12 +214,41 @@ public void reconnectToNodes(DiscoveryNodes discoveryNodes, Runnable onCompletio
+     * and disconnect-causing exception, if any
+     */
+    record DisconnectionHistory(long disconnectTimeMillis, @Nullable Exception disconnectCause) {
+        public long getDisconnectTimeMillis() {
+            return disconnectTimeMillis;
+        }
+
+        public Exception getDisconnectCause() {
+            return disconnectCause;
+        }
```
Nit: These methods do not seem necessary for a record class?

schase-es replied:
Ah, thanks for this -- I did look up making record fields public after accessing them from the test, but for some reason came away thinking I needed to write my own accessors. This helped me realize that I can just do `disconnectTimeMillis()` instead...

nicktindall reviewed (approved):
LGTM

schase-es committed:
Addressed final review feedback:
- fixed comment typo in ConnectionChangeListener
- removed hand-written accessors for DisconnectionHistory for public defaults
- corrected settings access for TimeValue/millis

schase-es committed:
Merge branch 'main' into ES-11448_log-network-disruption-reconnects

schase-es committed:
Merge branch 'main' into ES-11448_log-network-disruption-reconnects

schase-es committed:
Merge branch 'main' into ES-11448_log-network-disruption-reconnects
